core/src/main/java/io/spine/core/EventContextMixin.java

```diff
@@ -89,6 +90,39 @@ default ActorContext actorContext() {
         return actorContext;
     }
 
+    /**
+     * Obtains the ID of the first signal in a chain.
+     *
+     * <p>The root message is typically a {@code Command}. It can be an {@code Event} if the event
+     * was emitted by an actor directly, i.e. it was imported into the system.
+     *
+     * <p>If the associated message itself is the root of its chain, the ID cannot be assembled and
+     * thus an {@code Optional.empty()} is returned.
+     *
+     * @see Event#rootMessage()
+     */
+    @SuppressWarnings("deprecation") // For backward compatibility.
+    default Optional<MessageId> rootMessage() {
+        EventContext.OriginCase origin = getOriginCase();
+        switch (origin) {
+            case PAST_MESSAGE:
+                return Optional.of(getPastMessage().root());
+            case IMPORT_CONTEXT:
+                return Optional.empty();
+            case EVENT_CONTEXT:
+            case COMMAND_CONTEXT:
+            case ORIGIN_NOT_SET:
+            default:
+                @SuppressWarnings("DuplicateStringLiteralInspection") // Coincidence.
+                MessageId id = MessageId
```

Comment:
In the docs, you state that you'll return `empty` for root messages themselves. I assume that only the root message should have the `origin` as `NOT_SET`, right?